### DISCUSSION TITLE
Keyword driven framework approach in cypress

### DISCUSSION THREAD

**Discussion by shivani-saluja** [cypress-io/cypress#19665] [Questions and Help]
Hi All,

I was wondering if cypress supports keyword driven approach wherein all keywords and locators  and test steps are maintained in external file(excel sheet or json) and we may be able to build up a very generic framework so that it can be reused across multiple applications. Similar to the approach we have in Selenium.

Any help is much appreciated.

**Comment by dsgon** [comment]
Hi @shivani-saluja 


I think that Commands is the approach, because you can encapsulate actions/methods in commands that you can use or call into the test.

Example:

**login_actions.js**

```
function  login (username.password) {
    cy.get('username').should('be.visible').type(username)
    cy.get('password').should('be.visible').type(password)
    cy.get('login_btn').should('be.visible').click()
}
```

**commands**
import login module

```
Cypress.Commands.add('login',(username,password) => {
    login_module.login(username,password)
})
```

**test**

```
context('using commands as keyworddriven', () => ){
    describe('Test Suite for Commands as KD', () => {
        it(Test using commands as KD', () => {
            cy.login('username_value', 'password_value')
        })
    })
})
```

I hope this could help you

**Reply by j1000** [reply]
Good example. Cypress users have the flexibility of JavaScript and Node.js at their disposal (plus the handy built-in `cy.` commands), which effectively means they can utilize whatever organizational tools they can dream up.

It must be interesting writing tests for multiple applications at once. I find it challenging enough to keep track of one!
